### PR TITLE
Implement `IterativePhaseEstimation`

### PR DESCRIPTION
**Pull request by jlapeyre**
Add ~~`IPhaseEstimation`~~  `IterativePhaseEstimation` iterative phase estimation algorithm.

This is the plain-vanilla iterative phase estimation algorithm. It estimates the eigenphase of a pair (unitary, state).

### CODE REVIEW

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (Cryoris) [inline qiskit/algorithms/phase_estimators/iphase_estimation.py:3]: Could you update the copyrights in the new files to 2021?
```suggestion
# (C) Copyright IBM 2021.
```

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (Cryoris) [inline qiskit/algorithms/phase_estimators/iphase_estimation.py:27]: Can we call these algorithms `IterativePhaseEstimation` to be in sync with `IterativeAmplitudeEstimation`?

Comment (Cryoris) [inline qiskit/algorithms/phase_estimators/iphase_estimation.py:27]: This should also inherit from the `PhaseEstimator` interface, right?

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (Cryoris) [inline qiskit/algorithms/phase_estimators/iphase_estimation.py:37]: Can you move `num_iterations` from `estimate` into the initializer, since it's an algorithmic setting and not a problem setting?

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (Cryoris) [inline qiskit/algorithms/phase_estimators/iphase_estimation.py:48]: For consistency with `PhaseEstimation` can we explicitly pass the unitary etc instead of a problem dict? If we want to use an object to wrap the problem information I think we should introduce an analogous class to the `EstimationProblem` in amplitude estimation that is also used in the other phase estimation algorithms.

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (Cryoris) [inline qiskit/algorithms/phase_estimators/iphase_estimation.py:127]: I don't think we have to cover this case, the measurement will never contain anything else than 0 or 1, right?

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (Cryoris) [inline qiskit/algorithms/phase_estimators/iphase_estimation.py:165]: Should this derive from  the `PhaseEstimatorResult` (the interface, not PhaseEstimation's result class)?

File: test/python/algorithms/test_iphase_estimator.py
Comment (Cryoris) [inline test/python/algorithms/test_iphase_estimator.py:25]: Could you apply the same updates to these tests as we did in `TestPhaseEstimators`, namely
* default args should not be function calls (but None and then set a default in the test)
* group tests using `@ddt` and `@data`

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (Cryoris) [inline qiskit/algorithms/phase_estimators/iphase_estimation.py:137]: Should we add a check that the iterations are larger than 0? (Plus a test checking that)

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (jlapeyre) [reply qiskit/algorithms/phase_estimators/iphase_estimation.py:27]: Yes

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (jlapeyre) [reply qiskit/algorithms/phase_estimators/iphase_estimation.py:127]: Yes

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (jlapeyre) [reply qiskit/algorithms/phase_estimators/iphase_estimation.py]: It really looks like they have a different interface to me. One issue is that, in the parallel algorithm, you specify the number of qubits in the unitary. In the iterative, you pass the number of "iterations". These are not interchangeable.

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (jlapeyre) [reply qiskit/algorithms/phase_estimators/iphase_estimation.py]: done

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (jlapeyre) [reply qiskit/algorithms/phase_estimators/iphase_estimation.py]: done

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (jlapeyre) [reply qiskit/algorithms/phase_estimators/iphase_estimation.py]: Done

File: test/python/algorithms/test_iphase_estimator.py
Comment (jlapeyre) [reply test/python/algorithms/test_iphase_estimator.py]: done

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (jlapeyre) [reply qiskit/algorithms/phase_estimators/iphase_estimation.py]: The most important method of `PhaseEstimatorResult` is `most_likely_phase`. The most important method of `IterativePhaseEstimationResult` is `phase`. In the latter case, there is only one phase to return. Both methods return one phase, but the meaning of the phase returned is different. I suppose you could change the `PhaseEstimatorResult` method to `phase`, with the understanding that you are returning a single "best" phase, and it  is up to the implementation to decide how "best" is determined. In one case it would be the most likely phase. In the other, it would be the only phase.

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (Cryoris) [reply qiskit/algorithms/phase_estimators/iphase_estimation.py]: The interface is independent of the initializer, it just specifies that the algorithms can solve the same problem. So if the algorithms have the same interface I know I could do
```python
unitary = # some circuit
state_in = # some circuit

pe = PhaseEstimation(...)  # PE specific args
ipe = IterativePhaseEstimation(...)  # IPE specific args

result_pe = pe.estimate(unitary, state_in)   # same interface -> same "run" signature
result_ipe = ipe.estimate(unitary, state_in)
```
And PE and IPE solve the same problem, right?

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (Cryoris) [reply qiskit/algorithms/phase_estimators/iphase_estimation.py]: I see the point, however I think we should be able to use PE and IPE interchangeably, since they both solve the same problem. Can we find a name that is suitable for both? Maybe call both just `phase` and be clear in the docstring how that phase is determined?

**Comment by jlapeyre** [comment]
This PR is blocked (or I intend so) on #6238
After #6238 is merged, I'll push appropriate commits here.

File: test/python/algorithms/test_ipe.py
Comment (Cryoris) [inline test/python/algorithms/test_ipe.py:26]: Would it make sense to merge this with the PE tests? It looks like they are the same except you swap out the classes. For Amplitude Estimation we have that same structure, you could check there to compare 🙂

**Review by Cryoris — COMMENTED** [review]
Overall looks good! Some comments mainly on the tests 👍🏼

File: qiskit/algorithms/phase_estimators/ipe.py
Comment (jlapeyre) [reply qiskit/algorithms/phase_estimators/ipe.py]: done

File: test/python/algorithms/test_ipe.py
Comment (jlapeyre) [reply test/python/algorithms/test_ipe.py]: done

File: qiskit/algorithms/phase_estimators/ipe.py
Comment (jlapeyre) [reply qiskit/algorithms/phase_estimators/ipe.py]: done

File: test/python/algorithms/test_ipe.py
Comment (jlapeyre) [reply test/python/algorithms/test_ipe.py:26]: That makes some sense. TLDR: I'll try it.
OTOH there are two viewpoints: 1. "refactor relentlessly". And (something you can see whole talks on) 2. Don't refactor too early. Because, sometimes you want to add a class or change something, in which case you can add workarounds making your refactored design more brittle, or else un-refactor ("expand"?). In the current case the amount of code is small enough that it doesn't make much difference either way.

Comment (jlapeyre) [reply test/python/algorithms/test_ipe.py]: Done

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (jlapeyre) [reply qiskit/algorithms/phase_estimators/iphase_estimation.py]: Done

File: qiskit/algorithms/phase_estimators/iphase_estimation.py
Comment (jlapeyre) [reply qiskit/algorithms/phase_estimators/iphase_estimation.py]: Done

**Review by Cryoris — APPROVED** [review]
LGTM, thanks for implementing the changes! Since this has been stale for a while, I won't merge it before you had another look @jlapeyre 🙂

**Pull request merged**
